### DISCUSSION TITLE
Add per-component layout configuration

### DISCUSSION THREAD

**Discussion by calebporzio** [livewire/livewire#3138] [7. Bugs (archived)]
This is a feature spec:

```php
class Foo extends Component
{
    protected $layout = ['layouts.dashboard', 'body'];
    // or
    protected $layout = ['view' => 'layouts.dashboard', 'slot' => 'body'];

    protected $extends = ['layouts.dashboard', 'body'];
    // or
    protected $extends = ['view' => 'layouts.dashboard', 'section' => 'body'];

    // Also add support for methods:
    protected function layout()
    {
        return [...];
    }

    protected function extends()
    {
        return [...];
    }
}
```

This will allow apps with multiple different layouts to be able to extract the configuration from individual component's `render()` method into whatever abstraction they choose.

For example, you could extract a trait called `use WithDashboardLayout`, or even a parent component like `extends DashboardComponent` that has the layout settings.

**Comment by calebporzio** [comment]
I don't want to support setting the layout at the route level like Livewire V1, so this is an alternative to that solution that should provide mostly the same benefits.

**Comment by jszobody** [comment]
What if this isn't about making it like Livewire v1, and more about being consistent with the vanilla Laravel routing features?

If this was something Livewire v1 had invented, I get it. Let's do it better. But isn't there a lot of value to leveraging as much of the built-in Laravel routing features as possible, and making the transition from normal blade views to LW components that much simpler?

**Comment by calebporzio** [comment]
Please clarify @jszobody 

Laravel doesn't allow you to set any default layout for a blade view from a route.

Interested to know what you mean. Sample code would be great.

**Comment by iksaku** [comment]
The implementation of this spec will help for sure! It would help me reduce "boilerplate" defining the layout and section (or slot) for every "dashboard" component page (for example).

Totally into this :+1:

**Comment by joshhanley** [comment]
The other thing I think would be handy to support, would be not specifying a layout in the component class at all.

Since blade components, all my blade files have `<x-layouts.app></x-layouts.app>` tags wrapping the content in the .blade.php file and it would be nice to do the same with Livewire.

Not sure if that's even possible or will work though, but would be nice!

**Comment by iksaku** [comment]
@joshhanley yeah, that would be the nicest way, but due to the way Livewire works for component reactivity, it is not possible ATM, unless we use some kind of custom blade diffing to identify the root view part to replace 🤔

**Comment by joshhanley** [comment]
@iksaku yeah exactly, just thought I'd share the hopes and dreams 😁

**Comment by adamthehutt** [comment]
Another option (not mutually exclusive) would be allowing a config setting to configure the default layout for full-page components. I've got a PR ready to go for this if you like the idea: https://github.com/livewire/livewire/discussions/2221

**Comment by joshhanley** [comment]
Just to make visible here what @adamthehutt has done, so you don't need to open the PR:

He has suggested adding a key `'fullpage_layout' => 'layouts.app` to the livewire config file, and then does a check for the layout config in the invoke method of the component:

```php
'view' => $this->initialLayoutConfiguration['view'] ?? config('livewire.fullpage_layout', 'layouts.app'),
```

I personally think it's a great option, simple and effective, and you then don't need to add anything extra to each component manually.

I'd suggest though probably calling the key 'layout' or 'default_layout'.

**Comment by adamthehutt** [comment]
@joshhanley you're probably right about the config key. I was thinking it should specify that it's for full page components, but since the issue only applies to full page components (and "layouts" aren't really even relevant to other components) it's probably clear enough without the "fullpage" qualifier.

**Comment by iksaku** [comment]
Is there a way to automatically differentiate between Full Page Components and Partial Components?

IIRC, there’s none, and if so, then there should also be a way to tell your component “to be full page” and use that “default full page layout” in order to not collide with partial components.

I think that’s why the very issue suggests that syntax.

**Comment by joshhanley** [comment]
@iksaku the layout file is only applied in the __invoke method of a component, which is only called if instantiated by the router. I believe all child components are newed up but not invoked. Hence why the current layout default of `layouts.app` doesn't get applied to all components, just the full page ones.

That's why I think the config option is a good idea, as all it changes is swapping the hard coded `layouts.app` for a configurable option that falls back to `layouts.app`. Nothing else needs to change. 🙂

**Comment by iksaku** [comment]
@joshhanley That's true! I forgot about it... Then, having a default layout in config is a step forward 🥳

**Comment by joshhanley** [comment]
@adamthehutt I think you should submit a PR with your suggestion. 

I also had a thought for those that might be using the old layout extends instead of components:
- add a config option for 'layoutType' which can be 'extends' or 'component'. 
- Or another option 'useExtendsLayout` and is false by default, can be set to true.

 Have a look at Component.php:51 where you can set it on the component or has a default of 'component', that default could be a config option as well.

**Comment by adamthehutt** [comment]
@joshhanley PR submitted. #2246

**Comment by tomshaw** [comment]
What's the solution for users with multiple layouts?

**Reply by zepfietje** [reply]
Check out https://github.com/livewire/livewire/discussions/4195#discussioncomment-1628143.
I'm currently cooking up a pull request. 👨‍🍳

**Comment by baoanhng** [comment]
So this never made it through now I'm stucking with migrating many Route::layout from Livewire v1 to v2.

Edit: So I figured out you can make an abstract Component class and modify livewire.layout config in constructor. I think this should be enough.

**Comment by tomshaw** [comment]
I personally would prefer both props/methods to set a components layout/view. A data property would also be convenient instead of pipelining tons of component variables thru the render function something that looks ad hoc at best.
